manuals/pffexport.1
.Dd June 25, 2026
.Dt PFFEXPORT 1
.Os
.Sh NAME
.Nm pffexport
.Nd exports items stored in a Personal Folder File (OST, PAB and PST)
.Sh SYNOPSIS
.Nm pffexport
.Op Fl c Ar codepage
.Op Fl f Ar format
.Op Fl l Ar log_file
.Op Fl m Ar mode
.Op Fl t Ar target
.Op Fl dhqvV
.Ar source
.Sh DESCRIPTION
.Nm pffexport
is a utility to export items stored in a Personal Folder File (OST, PAB and PST)
.Pp
.Nm pffexport
is part of the
.Nm libpff
package.
.Nm libpff
is a library to access the Personal Folder File (OST, PAB and PST) format
.Pp
.Ar source
is the source file.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl c Ar codepage
codepage of ASCII strings, options: ascii, windows-874, windows-932, \
windows-936, windows-949, windows-950, windows-1250, windows-1251, windows-1252 \
(default), windows-1253, windows-1254, windows-1255, windows-1256, windows-1257 \
or windows-1258
.It Fl d
dumps the item values in a separate file: ItemValues.txt
.It Fl f Ar format
specify the preferred output format, options: all, html, rtf, text (default)
.It Fl h
shows this help
.It Fl l Ar log_file
specify the file in which to log information about the exported items
.It Fl m Ar mode
export mode, option: all, debug, items (default), recovered. 'all' exports the \
(allocated) items, orphan and recovered items. 'debug' exports all the \
(allocated) items, also those outside the root folder. 'items' exports the \
(allocated) items. 'recovered' exports the orphan and recovered items.
.It Fl q
quiet shows minimal status information
.It Fl t Ar target
specify the basename of the target directory to export to (default is the \
source filename) pffexport will add the following suffixes to the basename: \
\.export, .orphans, .recovered
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# pffexport file.pst
pffexport 20260625
.sp
Opening file.
Exporting items.
Exporting folder item 1 out of 1.
Exporting e-mail item 1 out of 1.
Exporting attachment 1 out of 3.
Exporting attachment 2 out of 3.
Exporting attachment 3 out of 3.
Export completed.
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh SEE ALSO
.Xr pffinfo 1
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libpff/issues
.Sh COPYRIGHT
Copyright (C) 2008-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO
libpff - clean up and refactoring

* refactor
** merge descriptors_index_tree and offsets_index_tree into index_tree?
** recover refactor libpff_recover_index_nodes to use index_node
** cache index nodes and/or index values
** cache local descriptors nodes and/or local descriptor values
** use or remove table->local_descriptors_cache

* libpff_table
  - move table header read functions into table header source file
  - move column definition read functions into column definition source file
  - cleaner reusable data structure? (libfdata)
  - create 7c table read entry definitions function
  - create ac table read entry definitions function
* have multi type work with data reference (in libpff_item) ?
* tree type implement use of initialize function (grep sizeof( libpff_tree_node_t ) )
* refactor direct use of array, list, and tree value attributes
* improve force decryption check
* refactor
  - create generic solution for data block handling for: individual blocks,
    array blocks, local descriptor blocks, table blocks
  - refactor libpff_local_descriptor_node_read to deal with compressed data
  blocks. read from data block?
    - fix libpff_local_descriptors_read_node_entry
  - item tree: wrap direct access to descriptors_index
  - recover: wrap direct access to descriptors and offsets_index
  - implement libfdata_list_set_element_by_index_with_mapped_size
  - 4k the back pointer of the array data block is the data identifier
    not the identifier?
    check for: "libpff_data_block_read: mismatch in data identifier:"
  - add a generic method to read data blocks from the file
  - local descriptors: cache local descriptor nodes ?
  - fix libfdata_list_set_element_data
  - table refactor entries into stand-alone multi set array ?
  - table read should empty the record_sets_array on read ?
  - record entry handle different data types on demand ?
  - what about table block?
  - use fvalue to store record entry data ?
  - libpff_item_values_get_record_entry_by_type add generic STRING TYPE ?
* code clean up
  - refactor table index into fdata list? with table blocks
  - does table index contain full copy of table data ? if so refactor
* refactor libpff_message_get_entry_value_utf8_string to match naming schema

* add threads RW lock to item
* remove libpff_data_block_clone
* pypff fix get_root_item

IOError: pypff_file_get_item_type_object: unable to retrieve item type.
libpff_table_read: invalid table: 0 - missing data identifier.
libpff_item_values_read: unable to read table.
libpff_item_values_get_record_entry_by_type: unable to read item values.
libpff_item_determine_type: unable to retrieve record entry: 0x001a.
libpff_item_get_type: unable to determine item type.

* add specific entry value flag for string types?
* add libpff_item_values_get_record_entry_by_type specific for string types?
* API add
  - libpff_multi_value_get_value_16bit
  - libpff_multi_value_get_value_floating_point
  - rename sub message to message ?
libpff
  - have libpff_record_entry_get_entry_multi_value use libfvalue data handle
    instead?
* pypff
  - create attachment item
  - make sure record set index and record entry index are uint32
  - add definitions
  - add record entry value functions
    pypff_record_entry_get_value_data_size
    get_value_data_as_boolean
    get_value_data_as_datetime
  - add multi value functions
  - add iterator objects (+/-)
  - add time as integer functions
  - add function to return offset of item
  - add more functions
  - add name to id map support
  - add message get_attachments_items function
  - add message get_recipients_items function

* local descriptor node caching
  - local descriptor node cache on top of data block cache ?
  - add cache timestamp
* add IPM.Post support (more1/Me.pst)
* corrupted
  - mark file with incorrect encryption as corrupted
* recovery
  - read data block size while scanning, ignore mismatches
  - check if local descriptors are read as block
  - handle corrupt file header and scan file?
* debug print
  - free map
  - densitiy list
* 64-bit 4k pages support
  - libpff_free_map.c
  - did the table value reference change?
* fix/check
  - check creation of .export when -t is a directory
  - remove copy file descriptors
  - Change most proxy pointers to const ?
  - check if libpff_item_get_entry_value_by_utf8_name works
  - Add function to easily map a named entry type to it's numeric value
  - add libpff_message_get_number_of_recipients() convenience function?
* add get entry id by name function
* add by_name functions (check corresponding TODO)
* fdata stream rewrite
  - array uses it as list
* clean up error path of libpff_item_tree_create_node

* libfdata update
  - improve stream, how deal with compressed data
  - fix libfdata_tree_initialize
* libpff IO handle
  - move index_nodes_vector and index_nodes_cache to file
* libfmapi
  - add debug output for PidTagFlagStatus
  - add debug output for PidLidSideEffects
- add debug output for PidLidTaskMode
- add debug output for PidLidAddressBookProviderArrayType
* implement libcdata support
  - refactor list
  - move tree node get value into libpff_item_initialize
* pffexport/export handle:
  - handle MIME attachments
  - protect against duplicate name clashes
  - dump attachment item values
  - add support for LIBPFF_ITEM_TYPE_COMMON (IPM)
  - separate item file into separate structure
  - move message specific funtionality into message file ?
  - allow to set output formatting utf8 or utf16
* item file:
  - create function to write enum values to the item file
  - create function to write flag values to the item file
* code clean up
  - what's the status of table block ?
  - unused and goto
* libpff: UTF8/UTF16
  - libpff_message_get_plain_text_body_size
  - libpff_message_get_rtf_body

* what about debug export and error notifications

Change item API into table API
* deprecrate and replace by copy data functions:
  - libpff_item_get_entry_value
  - libpff_item_get_entry_value_by_utf8_name

pfftools/pffexport.c
/*
 * Extracts items from a Personal Folder File (OST, PAB and PST).
 *
 * Copyright (C) 2008-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <memory.h>
#include <narrow_string.h>
#include <system_string.h>
#include <types.h>
#include <wide_string.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "export_handle.h"
#include "log_handle.h"
#include "pffinput.h"
#include "pfftools_getopt.h"
#include "pfftools_libcerror.h"
#include "pfftools_libclocale.h"
#include "pfftools_libcnotify.h"
#include "pfftools_libcpath.h"
#include "pfftools_libpff.h"
#include "pfftools_output.h"
#include "pfftools_signal.h"
#include "pfftools_unused.h"

export_handle_t *pffexport_export_handle = NULL;
libpff_file_t *pffexport_file            = NULL;
int pffexport_abort                      = 0;

/* Signal handler for pffexport
 */
void pffexport_signal_handler(
      pfftools_signal_t signal PFFTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function   = "pffexport_signal_handler";

	PFFTOOLS_UNREFERENCED_PARAMETER( signal )

	pffexport_abort = 1;

	if( pffexport_export_handle != NULL )
	{
		if( export_handle_signal_abort(
		     pffexport_export_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal export handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	if( pffexport_file != NULL )
	{
		if( libpff_file_signal_abort(
		     pffexport_file,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal file to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description    = \
		"Use pffexport to export items stored in a Personal Folder File (OST, PAB and PST).";

	pfftools_option_t options[ ] = {
		{ 'c', "codepage", "codepage of ASCII strings, options: ascii, windows-874, windows-932, windows-936, windows-949, windows-950, windows-1250, windows-1251, windows-1252 (default), windows-1253, windows-1254, windows-1255, windows-1256, windows-1257 or windows-1258" },
		{ 'd', NULL, "dumps the item values in a separate file: ItemValues.txt" },
		{ 'f', "format", "preferred output format, options: all, html, rtf, text (default)" },
		{ 'h', NULL, "shows this help" },
		{ 'l', "log_file", "logs information about the exported items" },
		{ 'm', "mode", "export mode, option: all, debug, items (default), recovered. 'all' exports the (allocated) items, orphan and recovered items. 'debug' exports all the (allocated) items, also those outside the the root folder. 'items' exports the (allocated) items. 'recovered' exports the orphan and recovered items." },
		{ 'q', NULL, "quiet shows minimal status information" },
		{ 't', "target", "specify the basename of the target directory to export to (default is the source filename) pffexport will add the following suffixes to the basename: .export, .orphans, .recovered" },
		{ 'v', NULL, "verbose output to stderr" },
		{ 'V', NULL, "print version" },
		{ 0, "source", "the source file" },
	};
	system_character_t options_string[ 32 ];

	libcerror_error_t *error                           = NULL;
	log_handle_t *log_handle                           = NULL;
	system_character_t *log_filename                   = NULL;
	system_character_t *option_codepage                = NULL;
	system_character_t *option_export_mode             = NULL;
	system_character_t *option_preferred_export_format = NULL;
	system_character_t *option_target_path             = NULL;
	system_character_t *path_separator                 = NULL;
	system_character_t *source                         = NULL;
	char *program                                      = "pffexport";
	system_integer_t option                            = 0;
	size_t source_length                               = 0;
	uint8_t dump_item_values                           = 0;
	uint8_t print_status_information                   = 1;
	int number_of_options                              = (int) ( sizeof( options ) / sizeof( pfftools_option_t ) );
	int result                                         = 0;
	int verbose                                        = 0;

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "pfftools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( pfftools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	pfftools_output_version_fprint(
	 stdout,
	 program );

	if( pfftools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = pfftools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				pfftools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'c':
				option_codepage = optarg;

				break;

			case (system_integer_t) 'd':
				dump_item_values = 1;

				break;

			case (system_integer_t) 'f':
				option_preferred_export_format = optarg;

				break;

			case (system_integer_t) 'h':
				pfftools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'l':
				log_filename = optarg;

				break;

			case (system_integer_t) 'm':
				option_export_mode = optarg;

				break;

			case (system_integer_t) 'q':
				print_status_information = 0;

				break;

			case (system_integer_t) 't':
				option_target_path = optarg;

				break;

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				pfftools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source file.\n" );

		pfftools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind ];

	if( option_target_path == NULL )
	{
		source_length = system_string_length(
		                 source );

		path_separator = system_string_search_character_reverse(
		                  source,
		                  (system_character_t) LIBCPATH_SEPARATOR,
		                  source_length );

		if( path_separator == NULL )
		{
			path_separator = source;
		}
		else
		{
			path_separator++;
		}
		option_target_path = path_separator;
	}
	libcnotify_verbose_set(
	 verbose );
	libpff_notify_set_stream(
	 stderr,
	 NULL );
	libpff_notify_set_verbose(
	 verbose );

	if( export_handle_initialize(
	     &pffexport_export_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to create export handle.\n" );

		goto on_error;
	}
#if defined( __clang_analyzer__ )
	__builtin_assume( pffexport_export_handle != NULL );
#endif
	pffexport_export_handle->print_status_information = print_status_information;

	if( option_export_mode != NULL )
	{
		result = export_handle_set_export_mode(
			  pffexport_export_handle,
			  option_export_mode,
			  &error );

		if( result == -1 )
		{
			fprintf(
			 stderr,
			 "Unable to set export mode.\n" );

			goto on_error;
		}
		else if( result == 0 )
		{
			fprintf(
			 stderr,
			 "Unsupported export mode defaulting to: items.\n" );
		}
	}
	pffexport_export_handle->dump_item_values = dump_item_values;

	if( option_preferred_export_format != NULL )
	{
		result = export_handle_set_preferred_export_format(
			  pffexport_export_handle,
			  option_preferred_export_format,
			  &error );

		if( result == -1 )
		{
			fprintf(
			 stderr,
			 "Unable to set preferred export format.\n" );

			goto on_error;
		}
		else if( result == 0 )
		{
			fprintf(
			 stderr,
			 "Unsupported preferred export format defaulting to: text.\n" );
		}
	}
	if( option_codepage != NULL )
	{
		result = export_handle_set_ascii_codepage(
		          pffexport_export_handle,
		          option_codepage,
		          &error );

		if( result == -1 )
		{
			fprintf(
			 stderr,
			 "Unable to set ASCII codepage in export handle.\n" );

			goto on_error;
		}
		else if( result == 0 )
		{
			fprintf(
			 stderr,
			 "Unsupported ASCII codepage defaulting to: windows-1252.\n" );
		}
	}
	if( export_handle_set_target_path(
	     pffexport_export_handle,
	     option_target_path,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to set target path.\n" );

		goto on_error;
	}
	result = export_handle_create_items_export_path(
	          pffexport_export_handle,
	          &error );

	if( result == -1 )
	{
		fprintf(
		 stderr,
		 "Unable to create items export path.\n" );

		goto on_error;
	}
	else if( result == 0 )
	{
		fprintf(
		 stderr,
		 "%" PRIs_SYSTEM " already exists.\n",
		 pffexport_export_handle->items_export_path );

		goto on_error;
	}
	result = export_handle_create_orphans_export_path(
	          pffexport_export_handle,
	          &error );

	if( result == -1 )
	{
		fprintf(
		 stderr,
		 "Unable to create orphans export path.\n" );

		goto on_error;
	}
	else if( result == 0 )
	{
		fprintf(
		 stderr,
		 "%" PRIs_SYSTEM " already exists.\n",
		 pffexport_export_handle->orphans_export_path );

		goto on_error;
	}
	result = export_handle_create_recovered_export_path(
	          pffexport_export_handle,
	          &error );

	if( result == -1 )
	{
		fprintf(
		 stderr,
		 "Unable to create recovered export path.\n" );

		goto on_error;
	}
	else if( result == 0 )
	{
		fprintf(
		 stderr,
		 "%" PRIs_SYSTEM " already exists.\n",
		 pffexport_export_handle->recovered_export_path );

		goto on_error;
	}
	if( log_handle_initialize(
	     &log_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to create log handle.\n" );

		goto on_error;
	}
	if( log_handle_open(
	     log_handle,
	     log_filename,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open log file: %" PRIs_SYSTEM ".\n",
		 log_filename );

		goto on_error;
	}
	if( libpff_file_initialize(
	     &pffexport_file,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to create file.\n" );

		goto on_error;
	}
	if( libpff_file_set_ascii_codepage(
	     pffexport_file,
	     pffexport_export_handle->ascii_codepage,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to set ASCII codepage.\n" );

		goto on_error;
	}
	if( pfftools_signal_attach(
	     pffexport_signal_handler,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to attach signal handler.\n" );

		goto on_error;
	}
	fprintf(
	 stdout,
	 "Opening file.\n" );

#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
	if( libpff_file_open_wide(
	     pffexport_file,
	     source,
	     LIBPFF_OPEN_READ,
	     &error ) != 1 )
#else
	if( libpff_file_open(
	     pffexport_file,
	     source,
	     LIBPFF_OPEN_READ,
	     &error ) != 1 )
#endif
	{
		fprintf(
		 stderr,
		 "Error opening file: %" PRIs_SYSTEM ".\n",
		 source );

		goto on_error;
	}
	if( export_handle_export_file(
	     pffexport_export_handle,
	     pffexport_file,
	     log_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to export file.\n" );

		goto on_error;
	}
	if( libpff_file_close(
	     pffexport_file,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close file.\n" );

		goto on_error;
	}
	if( pfftools_signal_detach(
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to detach signal handler.\n" );

		goto on_error;
	}
	if( libpff_file_free(
	     &pffexport_file,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free file.\n" );

		goto on_error;
	}
	if( log_handle_close(
	     log_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close log file.\n" );

		goto on_error;
	}
	if( log_handle_free(
	     &log_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free log handle.\n" );

		goto on_error;
	}
	if( export_handle_free(
	     &pffexport_export_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free export handle.\n" );

		goto on_error;
	}
	if( pffexport_abort != 0 )
	{
		fprintf(
		 stdout,
		 "Export aborted.\n" );

		return( EXIT_FAILURE );
	}
	fprintf(
	 stdout,
	 "Export completed.\n" );

	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( pffexport_file != NULL )
	{
		libpff_file_close(
		 pffexport_file,
		 NULL );
		libpff_file_free(
		 &pffexport_file,
		 NULL );
	}
	if( log_handle != NULL )
	{
		log_handle_close(
		 log_handle,
		 NULL );
		log_handle_free(
		 &log_handle,
		 NULL );
	}
	if( pffexport_export_handle != NULL )
	{
		export_handle_free(
		 &pffexport_export_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

